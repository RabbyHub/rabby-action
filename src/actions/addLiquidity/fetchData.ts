import PQueue from 'p-queue';
import {
  AddLiquidityRequireData,
  FetchActionRequiredData,
  ParsedActionData,
} from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import BigNumber from 'bignumber.js';

export const fetchDataAddLiquidity: FetchActionRequiredData<{
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: any;
  token1: any;
}> = async (options, likeAction) => {
  const { actionData, apiProvider, sender, walletProvider, chainId } = options;

  const action =
    likeAction || (<ParsedActionData<'transaction'>>actionData).addLiquidity;

  if (!action || options.type === 'text') {
    return {};
  }

  const { receiver, exchange_rate, token0, token1 } = action;

  const fair_exchange_rate = new BigNumber(token0.price).div(token1.price);

  const diff = +new BigNumber(exchange_rate)
    .minus(fair_exchange_rate)
    .abs()
    .div(fair_exchange_rate)
    .times(100)
    .toFixed(4);

  const receiverInWallet = await walletProvider.hasAddress(receiver);

  const id =
    options.type === 'transaction'
      ? options.tx.to
      : options.actionData.contractId;
  if (!id) {
    return {};
  }

  const result: AddLiquidityRequireData = {
    id,
    protocol: null,
    bornAt: 0,
    rank: null,
    sender,
    receiverInWallet,
    hasInteraction: false,
    receiver,
    poolRate: exchange_rate,
    token0,
    token1,
    marketRate: fair_exchange_rate.toFixed(),
    diff,
  };
  const queue = new PQueue();

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(id, chainId);
    if (!contractInfo) {
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.bornAt = contractInfo.create_at;

      result.protocol = contractInfo.protocol;
    }
  });

  queue.add(async () => {
    const hasInteraction = await apiProvider.hasInteraction(
      sender,
      chainId,
      id
    );
    result.hasInteraction = hasInteraction.has_interaction;
  });

  await waitQueueFinished(queue);

  return result;
};
