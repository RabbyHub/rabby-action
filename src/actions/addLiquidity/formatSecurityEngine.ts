import {
  AddLiquidityRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineAddLiquidity: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId } = options;

    if (!actionData.addLiquidity || !chainId) {
      return {} as any;
    }

    const data = requireData as AddLiquidityRequireData;

    return {
      addLiquidity: {
        receiverMatch:
          data.receiver?.toLowerCase() === data.sender.toLowerCase(),
        diff: data.diff,
      },
    };
  };
