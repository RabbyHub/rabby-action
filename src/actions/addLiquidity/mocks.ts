export const preExecData = {
  abi: {
    func: 'mint',
    params: [
      [
        '0x18Bc5bcC660cf2B9cE3cd51a404aFe1a0cBD3C22',
        '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913',
        10,
        -4970,
        -4960,
        2955759909,
        885199302,
        2940981110,
        880773306,
        '0xE2866F2291B5c5bafD2858B9247f93745108A523',
        1754366820,
        0,
      ],
    ],
  },
  abi_str:
    '{\n    "func": "mint",\n    "params": [\n        [\n            "0x18Bc5bcC660cf2B9cE3cd51a404aFe1a0cBD3C22",\n            "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",\n            10,\n            -4970,\n            -4960,\n            2955759909,\n            885199302,\n            2940981110,\n            880773306,\n            "0xE2866F2291B5c5bafD2858B9247f93745108A523",\n            1754366820,\n            0\n        ]\n    ]\n}',
  balance_change: {
    success: false,
    error: {
      code: 1002,
      msg: 'execution reverted: execution reverted',
    },
    send_token_list: [],
    receive_token_list: [],
    send_nft_list: [],
    receive_nft_list: [],
    usd_value_change: 0,
  },
  gas: {
    success: true,
    error: null,
    gas_used: 29320,
    gas_limit: 0,
    gas_ratio: 1.3,
  },
  is_gnosis: false,
  native_token: {
    id: 'base',
    chain: 'base',
    name: 'ETH',
    symbol: 'ETH',
    display_symbol: null,
    optimized_symbol: 'ETH',
    decimals: 18,
    logo_url:
      'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
    protocol_id: '',
    price: 3629.55,
    price_24h_change: 0.018385589869839923,
    credit_score: 46503412.60907772,
    is_verified: true,
    is_scam: false,
    is_suspicious: null,
    is_core: true,
    is_wallet: true,
    time_at: null,
    low_credit_score: false,
    amount: 1e-5,
  },
  pre_exec_version: 'v2',
  pre_exec: {
    success: false,
    error: {
      code: 1002,
      msg: 'execution reverted: execution reverted',
    },
  },
  type_call: {
    action: 'mint',
    contract: '0x827922686190790b37229fd06084350e74485b72',
    contract_protocol_name: 'Aerodrome V3',
    contract_protocol_logo_url:
      'https://static.debank.com/image/project/logo_url/base_aerodrome/f02d753bc321dc8ba480f0424a686482.png',
  },
  trace_id: '663ae6c6c0ac4b1aa5651fadba316de2',
} as any;

export const parseTxData = {
  action: { type: 'cancel_tx' },
  contract_call: null,
  log_id: 441860918,
} as any;

export const txData = {
  chainId: 42161,
  from: '0x4686fd7c0bc5d06346d416808916ba1dbd3f2336',
  to: '0x4686fd7c0bc5d06346d416808916ba1dbd3f2336',
  value: '0x0',
  data: '0x',
  gas: '0x0',
  maxFeePerGas: '0xb71b00',
  maxPriorityFeePerGas: '0xb71b00',
  nonce: '0x4',
};
