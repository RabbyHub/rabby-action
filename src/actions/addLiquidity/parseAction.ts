import { ParseAction } from '../../types';
import { TokenItem } from '@rabby-wallet/rabby-api/dist/types';

interface AddLiquidityActionData {
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: TokenItem;
  token1: TokenItem;
}

export const parseActionAddLiquidity: ParseAction = (options) => {
  const { data } = options;

  if (data?.type !== 'add_liquidity') {
    return {};
  }

  const {
    receiver,
    exchange_rate,
    min_exchange_rate,
    max_exchange_rate,
    token0,
    token1,
  } = data.data as unknown as AddLiquidityActionData;

  return {
    addLiquidity: {
      receiver,
      exchange_rate,
      min_exchange_rate,
      max_exchange_rate,
      token0,
      token1,
    },
  };
};
