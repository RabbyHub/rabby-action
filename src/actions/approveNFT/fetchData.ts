import PQueue from 'p-queue';
import { ApproveNFTRequireData, FetchActionRequiredData } from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import { fetchHasInteraction } from '../../fetches/fetchHasInteraction';

export const fetchDataApproveNFT: FetchActionRequiredData<{
  spender: string;
}> = async (options, likeAction) => {
  const queue = new PQueue();
  const { sender, apiProvider, chainId, actionData, extraActionDataState } =
    options;
  const action = likeAction || actionData.approveNFT;

  if (!action) {
    return {};
  }

  const result: ApproveNFTRequireData = {
    isEOA: false,
    contract: null,
    riskExposure: 0,
    rank: null,
    bornAt: 0,
    protocol: null,
    isDanger: false,
    tokenBalance: '0',
    hasInteraction: null,
    extraState: {
      hasInteraction: () => undefined,
    },
  };

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(
      action.spender,
      chainId
    );
    if (!contractInfo) {
      result.isEOA = true;
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.riskExposure = contractInfo.top_nft_spend_usd_value;
      result.bornAt = contractInfo.create_at;
      result.isDanger =
        contractInfo.is_danger.auto || contractInfo.is_danger.edit;
      result.protocol = contractInfo.protocol;
    }

    if (result.isEOA) {
      queue.add(async () => {
        const { desc } = await apiProvider.addrDesc(action.spender);
        result.bornAt = desc.born_at;
      });
    }
  });

  fetchHasInteraction({
    apiProvider,
    sender,
    chainId,
    spender: action.spender,
    extraActionDataState,
    queue,
    result,
  });

  await waitQueueFinished(queue);
  return result;
};
