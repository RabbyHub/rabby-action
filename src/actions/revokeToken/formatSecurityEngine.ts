import { FormatSecurityEngineContext } from '../../types';

export const formatSecurityEngineRevokeToken: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId, provider } = options;

    if (!actionData.revokeToken) {
      return {};
    }

    const { gasUsed } = actionData.revokeToken;
    return {
      revokeApprove: {
        gasUsed: gasUsed || 0,
        chainId,
      },
    };
  };
