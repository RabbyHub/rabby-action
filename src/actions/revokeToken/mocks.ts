export const preExecData = {
  abi: {
    func: 'approve',
    params: ['0x816f722424B49Cf1275cc86DA9840Fbd5a6167e9', 0],
  },
  abi_str:
    '{\n    "func": "approve",\n    "params": [\n        "0x816f722424B49Cf1275cc86DA9840Fbd5a6167e9",\n        0\n    ]\n}',
  balance_change: {
    success: true,
    error: null,
    send_token_list: [],
    receive_token_list: [],
    send_nft_list: [],
    receive_nft_list: [],
    usd_value_change: 0,
  },
  gas: {
    success: true,
    error: null,
    gas_used: 37606,
    gas_limit: 43492,
    gas_ratio: 1.3,
  },
  is_gnosis: false,
  native_token: {
    id: 'arb',
    chain: 'arb',
    name: 'ETH',
    symbol: 'ETH',
    display_symbol: null,
    optimized_symbol: 'ETH',
    decimals: 18,
    logo_url:
      'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
    protocol_id: '',
    price: 2752.79,
    price_24h_change: -0.0023605601525843664,
    credit_score: 100000000.0,
    is_verified: true,
    is_scam: false,
    is_suspicious: false,
    is_core: true,
    is_wallet: true,
    time_at: 1622131200.0,
    amount: 8.941029391638e-6,
  },
  pre_exec_version: 'v2',
  pre_exec: {
    success: true,
    error: null,
  },
  type_cancel_token_approval: {
    spender: '0x816f722424b49cf1275cc86da9840fbd5a6167e9',
    spender_protocol_name: 'Orderly Bridge',
    spender_protocol_logo_url:
      'https://static.debank.com/image/project/logo_url/arb_orderly/a3d0c1dd679b7db6db038f405816f67d.png',
    token_symbol: 'USDC',
    is_nft: false,
    nft: null,
  },
  trace_id: '65512bb8b45f4efab9424f798c0548b3',
} as any;

export const parseTxData = {
  action: {
    type: 'revoke_token',
    data: {
      spender: '0x816f722424b49cf1275cc86da9840fbd5a6167e9',
      token: {
        id: '0xaf88d065e77c8cc2239327c5edb3a432268e5831',
        chain: 'arb',
        name: 'USD Coin',
        symbol: 'USDC',
        display_symbol: null,
        optimized_symbol: 'USDC',
        decimals: 6,
        logo_url:
          'https://static.debank.com/image/arb_token/logo_url/0xaf88d065e77c8cc2239327c5edb3a432268e5831/fffcd27b9efff5a86ab942084c05924d.png',
        protocol_id: '',
        price: 0.9996001599360256,
        price_24h_change: 0.0,
        credit_score: 13584005.874952497,
        is_verified: true,
        is_scam: false,
        is_suspicious: false,
        is_core: true,
        is_wallet: true,
        time_at: 1667248932.0,
        amount: 0,
        raw_amount: '0',
      },
    },
  },
  contract_call: {
    func: 'approve',
    contract: {
      id: '0xaf88d065e77c8cc2239327c5edb3a432268e5831',
      protocol: null,
    },
  },
  log_id: 87892138,
} as any;

export const txData = {
  chainId: 42161,
  data: '0x095ea7b3000000000000000000000000816f722424b49cf1275cc86da9840fbd5a6167e90000000000000000000000000000000000000000000000000000000000000000',
  from: '0x99bf13b64758d1c71dcb8a16897edae9e8c12039',
  gas: '',
  gasPrice: '0xc65d40',
  nonce: '0x103',
  to: '0xaf88d065e77c8cc2239327c5edb3a432268e5831',
  value: '0x0',
};
