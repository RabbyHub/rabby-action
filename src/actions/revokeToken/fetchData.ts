import { FetchActionRequiredData } from '../../types';
import { fetchDataApproveToken } from '../approveToken/fetchData';

export const fetchDataRevokeToken: FetchActionRequiredData = async (
  options
) => {
  if (!options.actionData.revokeToken) {
    return {};
  }
  return fetchDataApproveToken(options, options.actionData.revokeToken);
};
