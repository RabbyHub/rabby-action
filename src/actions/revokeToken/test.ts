import {
  apiProvider,
  ETH_CHAIN_ID,
  formatProvider,
  SENDER,
  walletProvider,
} from '../../../__mocks__';
import { ParsedTransactionActionData } from '../../types';
import { fetchDataRevokeToken } from './fetchData';
import { formatSecurityEngineRevokeToken } from './formatSecurityEngine';
import { parseTxData, preExecData, txData } from './mocks';
import { parseActionRevokeToken } from './parseAction';
import {
  parseAction,
  fetchActionRequiredData,
  formatSecurityEngineContext,
} from '../..';

/**
 * Rabby Revoke Token
 */
test.each([
  [
    parseActionRevokeToken,
    fetchDataRevokeToken,
    formatSecurityEngineRevokeToken,
  ],
  [parseAction, fetchActionRequiredData, formatSecurityEngineContext],
])('RevokeToken', async (_parseAction, _fetchData, _format) => {
  const actionData = _parseAction({
    type: 'transaction',
    data: parseTxData['action'],
    balanceChange: preExecData.balance_change,
    sender: txData.from,
    preExecVersion: preExecData.pre_exec_version,
    gasUsed: preExecData.gas.gas_used,
    tx: txData,
  }) as ParsedTransactionActionData;
  expect(actionData).toMatchSnapshot('parseActionRevokeToken');

  const requireData = await _fetchData({
    type: 'transaction',
    actionData,
    contractCall: parseTxData.contract_call,
    chainId: ETH_CHAIN_ID,
    sender: SENDER,
    walletProvider,
    tx: txData,
    apiProvider,
  });

  expect(requireData).toMatchSnapshot('fetchDataRevokeToken');

  const ctx = await _format({
    type: 'transaction',
    actionData,
    requireData,
    chainId: ETH_CHAIN_ID,
    isTestnet: false,
    provider: formatProvider,
  });
  expect(ctx).toMatchSnapshot('formatSecurityEngineRevokeToken');
});
