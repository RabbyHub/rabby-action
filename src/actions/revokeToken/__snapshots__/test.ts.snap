// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RevokeToken: fetchDataRevokeToken 1`] = `
{
  "bornAt": 1591388241,
  "contract": null,
  "isDanger": false,
  "isEOA": false,
  "protocol": {
    "id": "uniswap2",
    "logo_url": "https://static.debank.com/image/project/logo_url/uniswap2/87a541b3b83b041c8d12119e5a0d19f0.png",
    "name": "Uniswap V2",
  },
  "rank": 65,
  "riskExposure": 20000000000,
  "token": {
    "amount": 7.625238,
    "chain": "eth",
    "credit_score": 18127653.217581443,
    "decimals": 6,
    "display_symbol": null,
    "id": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "is_core": true,
    "is_scam": false,
    "is_suspicious": false,
    "is_verified": true,
    "is_wallet": true,
    "logo_url": "https://static.debank.com/image/eth_token/logo_url/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48/fffcd27b9efff5a86ab942084c05924d.png",
    "name": "USD Coin",
    "optimized_symbol": "USDC",
    "price": 0.9998000399920016,
    "price_24h_change": 0.00009998000399924235,
    "protocol_id": "",
    "raw_amount": 7625238,
    "raw_amount_hex_str": "0x745a16",
    "symbol": "USDC",
    "time_at": 1533324504,
  },
}
`;

exports[`RevokeToken: fetchDataRevokeToken 2`] = `
{
  "bornAt": 1591388241,
  "contract": null,
  "isDanger": false,
  "isEOA": false,
  "protocol": {
    "id": "uniswap2",
    "logo_url": "https://static.debank.com/image/project/logo_url/uniswap2/87a541b3b83b041c8d12119e5a0d19f0.png",
    "name": "Uniswap V2",
  },
  "rank": 65,
  "riskExposure": 20000000000,
  "token": {
    "amount": 7.625238,
    "chain": "eth",
    "credit_score": 18127653.217581443,
    "decimals": 6,
    "display_symbol": null,
    "id": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "is_core": true,
    "is_scam": false,
    "is_suspicious": false,
    "is_verified": true,
    "is_wallet": true,
    "logo_url": "https://static.debank.com/image/eth_token/logo_url/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48/fffcd27b9efff5a86ab942084c05924d.png",
    "name": "USD Coin",
    "optimized_symbol": "USDC",
    "price": 0.9998000399920016,
    "price_24h_change": 0.00009998000399924235,
    "protocol_id": "",
    "raw_amount": 7625238,
    "raw_amount_hex_str": "0x745a16",
    "symbol": "USDC",
    "time_at": 1533324504,
  },
}
`;

exports[`RevokeToken: formatSecurityEngineRevokeToken 1`] = `
{
  "revokeApprove": {
    "chainId": "eth",
    "gasUsed": 37606,
  },
}
`;

exports[`RevokeToken: formatSecurityEngineRevokeToken 2`] = `
{
  "revokeApprove": {
    "chainId": "eth",
    "gasUsed": 37606,
  },
}
`;

exports[`RevokeToken: parseActionRevokeToken 1`] = `
{
  "revokeToken": {
    "gasUsed": 37606,
    "spender": "0x816f722424b49cf1275cc86da9840fbd5a6167e9",
    "token": {
      "amount": 0,
      "chain": "arb",
      "credit_score": 13584005.874952497,
      "decimals": 6,
      "display_symbol": null,
      "id": "0xaf88d065e77c8cc2239327c5edb3a432268e5831",
      "is_core": true,
      "is_scam": false,
      "is_suspicious": false,
      "is_verified": true,
      "is_wallet": true,
      "logo_url": "https://static.debank.com/image/arb_token/logo_url/0xaf88d065e77c8cc2239327c5edb3a432268e5831/fffcd27b9efff5a86ab942084c05924d.png",
      "name": "USD Coin",
      "optimized_symbol": "USDC",
      "price": 0.9996001599360256,
      "price_24h_change": 0,
      "protocol_id": "",
      "raw_amount": "0",
      "symbol": "USDC",
      "time_at": 1667248932,
    },
  },
}
`;

exports[`RevokeToken: parseActionRevokeToken 2`] = `
{
  "revokeToken": {
    "gasUsed": 37606,
    "spender": "0x816f722424b49cf1275cc86da9840fbd5a6167e9",
    "token": {
      "amount": 0,
      "chain": "arb",
      "credit_score": 13584005.874952497,
      "decimals": 6,
      "display_symbol": null,
      "id": "0xaf88d065e77c8cc2239327c5edb3a432268e5831",
      "is_core": true,
      "is_scam": false,
      "is_suspicious": false,
      "is_verified": true,
      "is_wallet": true,
      "logo_url": "https://static.debank.com/image/arb_token/logo_url/0xaf88d065e77c8cc2239327c5edb3a432268e5831/fffcd27b9efff5a86ab942084c05924d.png",
      "name": "USD Coin",
      "optimized_symbol": "USDC",
      "price": 0.9996001599360256,
      "price_24h_change": 0,
      "protocol_id": "",
      "raw_amount": "0",
      "symbol": "USDC",
      "time_at": 1667248932,
    },
  },
}
`;
